contracts/DependencyRegistryV1.sol

```diff
@@ -0,0 +1,1074 @@
+            return 0x0;
+        }
+
+        assembly {
+            result := mload(add(source, 32))
+        }
+    }
+
+    /**
+     * @notice Converts bytes to hexadecimal string representation.
+     * https://ethereum.stackexchange.com/a/126928
+     * @param buffer bytes to be converted.
+     * @return string representation of buffer.
+     */
+    function bytesToHexString(
+        bytes memory buffer
+    ) public pure returns (string memory) {
+        // Fixed buffer size for hexadecimal convertion
+        bytes memory converted = new bytes(buffer.length * 2);
+
+        bytes memory _base = "0123456789abcdef";
+
+        for (uint256 i = 0; i < buffer.length; i++) {
+            converted[i * 2] = _base[uint8(buffer[i]) / _base.length];
+            converted[i * 2 + 1] = _base[uint8(buffer[i]) % _base.length];
+        }
+
+        return string(abi.encodePacked("0x", converted));
+    }
+
+    function getTokenHtml(
+        address _contractAddress,
+        uint256 _tokenId
+    ) external view returns (string memory) {
+        uint256 ONE_MILLION = 1_000_000;
+
+        uint256 projectId = _tokenId / ONE_MILLION;
+        bytes32 dependencyType = stringToBytes32(
+            this.getDependencyTypeForProject(_contractAddress, projectId)
+        );
+        bytes32 tokenHash = IDependencyRegistryCompatibleV0(_contractAddress)
+            .tokenIdToHash(_tokenId);
+
+        WrappedScriptRequest[] memory requests = new WrappedScriptRequest[](5);
+
+        requests[0]
+            .scriptContent = 'let css="html{height:100%}body{min-height:100%;margin:0;padding:0}canvas{padding:0;margin:auto;display:block;position:absolute;top:0;bottom:0;left:0;right:0}",head=document.head,style=document.createElement("style");head.appendChild(style),style.type="text/css",style.appendChild(document.createTextNode(css));';
+
+        requests[1].scriptContent = abi.encodePacked(
+            'let tokenData = {"tokenId":"',
+            Strings.toString(_tokenId),
+            '"',
+            ',"hash":"',
+            bytesToHexString(abi.encodePacked(tokenHash)),
+            '"}'
+        );
+
+        requests[2].name = "dependency";
+        requests[2].contractData = abi.encode(dependencyType);
+        requests[2].contractAddress = address(this);
+        requests[2].wrapType = 2; // <script type="text/javascript+gzip" src="data:text/javascript;base64,[script]"></script>
+
+        requests[3].name = "gunzipScripts-0.0.1.js";
+        requests[3].wrapType = 1; // <script src="data:text/javascript;base64,[script]"></script>
+        requests[3].contractAddress = ethFsAddress;
+
+        requests[4].name = "project";
+        requests[4].contractData = abi.encode(_contractAddress, projectId);
+        requests[4].contractAddress = address(this);
+        requests[4].wrapType = 0; // <script>[script]</script>
+
+        uint256 bufferSize = scriptyBuilder.getBufferSizeForEncodedHTMLWrapped(
+            requests
+        );
```

Comment:
This here is a bit problematic. 

Just to give a little bit context.. scripty.sol gets bufferSize to allocate a big amount of container memory
which then is subsequently filled without needing to reallocate. This way scripty can stitch scripts together in a gas efficient way.

`scriptyBuilder.getBufferSizeForEncodedHTMLWrapped()` -> This method actually builds the HTML and then returns the buffer size of final HTML. So calling this method on-chain basically doubles the HTML building operation which might explain why you had gas issue initially. It's intended to be used mostly off-chain.

So I wondering if this approach would be better for AB:
- getDependencyScript(params..) -> String  - Returns full dependency script text. (Similar to getScript method but tuned for dependency)
- getArtScript(params..) -> String - Returns full art script. (Similar to getScript but tuned for art scripts)
- getArtControllerScript(params...) -> String returns the let tokenData = {"tokenData script.

Then you can get the final buffer size:
- dependencyScript.length + artScript.length + artControllerScript.length

and pass it to scripty.

These are few benefits I see:
- All components needed to build final HTML would be separated and can be called publicly in a clean way. 
- No extra contract call gas cost from scripty. (Right now registry calls scripty and then scripty calls registry back to reach the scripts.)

What do you think?

Reply:
Ah good call, that makes a lot of sense!